spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -125,12 +132,29 @@ private[cassandra] class CassandraSourceRelation(
     finalPushdown
   }
 
+  private def isOptimizationAvailable(optimizedFilters: List[Array[Filter]]): Boolean =
+    enableWhereClauseOptimization && optimizedFilters.size > 1 &&
+      optimizedFilters.sliding(2).forall{ set =>
+        // check whether all non-pushed down filters are equals for each separate rdd
+        predicatePushDown(set.head).handledBySpark == predicatePushDown(set.last).handledBySpark
+      }
+  
+
   override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
     val prunedRdd = maybeSelect(baseRdd, requiredColumns)
     val prunedFilteredRdd = {
       if(filterPushdown) {
-        val pushdownFilters = predicatePushDown(filters).handledByCassandra.toArray
-        val filteredRdd = maybePushdownFilters(prunedRdd, pushdownFilters)
+        val optimizedFilters = new FiltersOptimizer(filters).build()
+        val optimizationCanBeApplied = isOptimizationAvailable(optimizedFilters)
+        val filteredRdd = if(optimizationCanBeApplied) {
```

Comment:
I think this is a rather dangerous optimization sometimes, so I think we should default to off. For example

Table where x < 3 or x > 5 and x ranges from 1 to 10000. Doing two scans here is probably much more expensive than a single scan.